### PR TITLE
Acc 3047 add opa bundle endpoint

### PR DESCRIPTION
**Pull request by AndreasVAmexio**

### CODE REVIEW

File: contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    private final RegoPolicyLoader policyLoader;
+    private final OpaBundleBuilder bundleBuilder = new OpaBundleBuilder();
+    private final String policyPackage;
+
+    private final Object lock = new Object[0];
+
+    // Using volatile is fine here, since the properties of the bundle object are never updated.
+    // The object as a whole is set.
+    @SuppressWarnings("java:S3077")
+    private volatile PolicyBundle bundle;
```
Comment (vierbergenlars) [inline contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java]: This keeps the bundle permanently in memory, even though in the normal case we never use it again after OPA has loaded the bundle once (barring the exceptional circumstance when the OPA container is restarted).

It is fine to re-generate the bundle when we need it.

File: contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            }
+        }
+        return bundle;
+    }
+
+    private PolicyBundle build()
+            throws IOException, BlueprintArtifactException, BlueprintArtifactItemUnreadableException {
+        var policy = policyLoader.readPolicy(policyPackage);
+        var content = bundleBuilder.build(policy, policyPackage);
+        var etag = digest(content);
```
Comment (vierbergenlars) [inline contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java]: The `ETag` of the bundle is calculated from the content here. As we know the bundle can not change after the fact, this is not really necessary.
Also, currently you're caching the bundle after building it, so the `ETag` can not change at all.
It would be fine to work with a static value.

File: contentgrid-appserver-autoconfigure/src/main/java/com/contentgrid/appserver/autoconfigure/actuator/PolicyBundleEtagManagementContextConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * Registers {@link PolicyBundleEtagFilter} in front of the policy bundle endpoint.
+ * <p>
+ * This is a {@link ManagementContextConfiguration} rather than a regular auto-configuration because a management
+ * server running on its own port lives in a child context, which a filter registered in the main context would
+ * never reach.
+ */
+@ManagementContextConfiguration(proxyBeanMethods = false)
+@ConditionalOnWebApplication(type = Type.SERVLET)
+@ConditionalOnClass(PolicyBundleEtagFilter.class)
+public class PolicyBundleEtagManagementContextConfiguration {
```
Comment (vierbergenlars) [inline contentgrid-appserver-autoconfigure/src/main/java/com/contentgrid/appserver/autoconfigure/actuator/PolicyBundleEtagManagementContextConfiguration.java]: It looks like this configuration is not imported anywhere. Is that correct and expected?

File: contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java [hunk 4]
```diff
@@ -67,6 +77,29 @@ static void opaProperties(DynamicPropertyRegistry registry) {
                 () -> "http://%s:%d".formatted(opa.getHost(), opa.getMappedPort(8181)));
     }
 
+    @BeforeAll
+    static void pushPolicyToOpa() throws Exception {
```
Comment (vierbergenlars) [inline contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java]: Why not let OPA pull the bundle instead?

File: contentgrid-appserver-actuators/src/test/java/com/contentgrid/appserver/actuator/ActuatorConfigurationTest.java [hunk 5]
```diff
@@ -66,6 +73,29 @@ void policyEndpointIsPublic() {
...
+
+        var entries = new LinkedHashMap<String, String>();
+        try (var archive = new TarArchiveInputStream(new GzipCompressorInputStream(new ByteArrayInputStream(body)))) {
+            TarArchiveEntry entry;
+            while ((entry = archive.getNextEntry()) != null) {
+                entries.put(entry.getName(), new String(archive.readAllBytes(), StandardCharsets.UTF_8));
+            }
+        }
+
+        assertThat(entries).containsOnlyKeys(".manifest", "policy.rego");
```
Comment (vierbergenlars) [inline contentgrid-appserver-actuators/src/test/java/com/contentgrid/appserver/actuator/ActuatorConfigurationTest.java]: If you're not testing the contents, it doesn't really make sense to put it in a hashmap

**Review by vierbergenlars — CHANGES_REQUESTED** [review]
As a general remark, it looks a bit over-engineered with the code spread over 5 different classes.

I understand the need for a layer to build the rego policy, because that is used in 2 places.
I also understand there being a class that creates the bundle tar.gz, so it can be tested separately without going through the actuator as well.

I don't really understand the need for separating `PolicyBundleSource` & `OpaBundleBuilder` (and then an intermediate `PolicyBundle` record)

The need for a separate filter to do the ETags is also unfortunate, but I understand that you can't return headers from an actuator endpoint.

File: contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java (hunk 1 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java]: I kept the entire bundle in memory for now because the filter is the first one to call the `load()` method. So the `PolicyBundleSource` needs to keep the build bundle in memory so the actuator endpoint can return it. If the bundle source doesn't cache the bundle, then every first request builds the bundle twice.
I'll look at the static etag suggestion first, that should solve this problem as well.

File: contentgrid-appserver-autoconfigure/src/main/java/com/contentgrid/appserver/autoconfigure/actuator/PolicyBundleEtagManagementContextConfiguration.java (hunk 3 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-autoconfigure/src/main/java/com/contentgrid/appserver/autoconfigure/actuator/PolicyBundleEtagManagementContextConfiguration.java]: It's imported by `org.springframework.boot.actuate.autoconfigure.web.ManagementContextConfiguration.imports`

File: contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java (hunk 4 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java]: I didn't go for pulling because I thought it would slow the test down unnecessarily. But the tests are fast enough with the opa polling, so I've updated them.

File: contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java (hunk 1 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java]: The static etag made the `PolicyBundle` caching unnecessary.

File: contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java (hunk 2 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-actuators/src/main/java/com/contentgrid/appserver/actuator/policy/PolicyBundleSource.java]: The `ETag` is now a random number generated at startup.

**Comment by AndreasVAmexio** [comment]
> As a general remark, it looks a bit over-engineered with the code spread over 5 different classes.
> 
> ...

I've refactored the code to use a random UUID generated at startup for the `ETag`. This allows the filter to be the sole owner of the etag and reduced the amount of classes.

File: contentgrid-appserver-actuators/src/test/java/com/contentgrid/appserver/actuator/ActuatorConfigurationTest.java [hunk 6]
```diff
@@ -66,6 +66,19 @@ void policyEndpointIsPublic() {
...
     }
 
+    @Test
+    void policyBundleEndpointIsPublic() {
+        var body = rest.get().uri("http://localhost:" + managementPort + "/actuator/policybundle")
+                .exchange()
+                .expectStatus().is2xxSuccessful()
+                .expectHeader().valueEquals("Content-Type", "application/gzip")
+                .expectBody(byte[].class)
+                .returnResult()
```
Comment (vierbergenlars) [inline contentgrid-appserver-actuators/src/test/java/com/contentgrid/appserver/actuator/ActuatorConfigurationTest.java]: Can you add a test here that the ETag is properly returned, and that making a second request with the ETag in `If-None-Match` will result in a 304 Not Modified response?

File: contentgrid-appserver-autoconfigure/src/main/java/com/contentgrid/appserver/autoconfigure/actuator/PolicyBundleEtagManagementContextConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * Runs after the spring security filter chain, so that a {@code 304} is never returned
+     * to a caller that security would have rejected.
+     */
+    private static final int FILTER_ORDER = DEFAULT_FILTER_ORDER + 10;
+
+    @Bean
+    @ConditionalOnBean(PolicyBundleActuator.class)
```
Comment (vierbergenlars) [inline contentgrid-appserver-autoconfigure/src/main/java/com/contentgrid/appserver/autoconfigure/actuator/PolicyBundleEtagManagementContextConfiguration.java]: I don't think you want to make this conditional on that bean.

This is not an autoconfiguration, so I don't know what the ordering is exactly compared to the registration of the `PolicyBundleActuator` bean.

Since there already is a check below to not register the filter when the endpoint path does not exist, I think it is fine to unconditionally create this bean.

File: contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java [hunk 8]
```diff
@@ -53,18 +63,61 @@ class OpaResidualAuthorizationTest {
 
     private static final String JWT_ISSUER = "https://tenant.example.com";
 
+    /**
+     * Fixed so it can be handed to OPA before the application binds it. OPA retries its bundle poll, so it
+     * tolerates the port not listening yet when the container starts.
+     */
+    private static final int MANAGEMENT_PORT = freePort();
```
Comment (vierbergenlars) [inline contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java]: You can actually set `management.server.port=0` in the configuration and then get it injected in a property.

That is already done in other tests, like here: https://github.com/xenit-eu/contentgrid-appserver/blob/98ff6687bf81f9e9ef40505b592e86243917ef3b/contentgrid-appserver-app/src/test/java/com/contentgrid/appserver/example/ContentgridPrometheusActuatorTest.java#L23

You can still call `Testcontainers.exposeHostPorts()` in a `@BeforeEach` method

File: contentgrid-appserver-actuators/src/test/java/com/contentgrid/appserver/actuator/ActuatorConfigurationTest.java (hunk 6 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-actuators/src/test/java/com/contentgrid/appserver/actuator/ActuatorConfigurationTest.java]: The `Etag` behaviour is already tested in `PolicyBundleEndpointTest.java` integration test. I didn't want to bloat the `ActuatorConfigurationTest` with detailed tests for each endpoint.

File: contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java (hunk 8 above)
Comment (AndreasVAmexio) [reply contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java]: I've tried the `@Value("${local.management.port}")`, but I couldn't get it working for this test. The opa test container needs the `managementPort` to know here to poll from. And the container needs to exist before the application startup so the `opa.service.url` property can be registered. Both need to happen statically, but the `@Value` doesn't work when the field is static. 
The only way I see `@Value` working is by manually managing the container lifecycle for each test, which doesn't seem worth it.

**Review by vierbergenlars — DISMISSED** [review]

File: contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java (hunk 8 above)
Comment (vierbergenlars) [reply contentgrid-appserver-integration-test/src/test/java/com/contentgrid/appserver/integration/test/opa/OpaResidualAuthorizationTest.java]: Hmm, that indeed does sound like a chicken-egg problem then

**Review by vierbergenlars — APPROVED** [review]
